File: frontend/src/types/log.ts
// Trade log types - aligned with backend API

import { type Period } from './chart';

// Pattern type from backend PatternEnum
export type LogPatternType =
  | 'head_and_shoulders_top'
  | 'head_and_shoulders_bottom'
  | 'double_top'
  | 'double_bottom'
  | 'triple_top'
  | 'triple_bottom'
  | 'triangle'
  | 'flag';

// Pattern type display names in Chinese
export const LOG_PATTERN_TYPE_LABELS: Record<LogPatternType, string> = {
  head_and_shoulders_top: '头肩顶',
  head_and_shoulders_bottom: '头肩底',
  double_top: '双顶',
  double_bottom: '双底',
  triple_top: '三重顶',
  triple_bottom: '三重底',
  triangle: '三角形',
  flag: '旗形',
};

export interface TradeLog {
  id: string;
  stockCode: string;
  stockName?: string;
  period: Period;
  patternType?: LogPatternType;
  entryPrice: number;
  stopLoss?: number;
  takeProfit?: number;
  exitPrice?: number;
  quantity: number;
  profitLoss?: number;
  notes?: string;
  tags: string[];
  tradeTime?: string;
  createdAt: string;
  updatedAt: string;
}

export interface TradeLogCreate {
  stockCode: string;
  stockName?: string;
  period: Period;
  patternType?: LogPatternType;
  entryPrice: number;
  stopLoss?: number;
  takeProfit?: number;
  exitPrice?: number;
  quantity: number;
  profitLoss?: number;
  notes?: string;
  tags?: string[];
  tradeTime?: string;
}

export interface TradeLogUpdate {
  stockCode?: string;
  stockName?: string;
  period?: Period;
  patternType?: LogPatternType;
  entryPrice?: number;
  stopLoss?: number;
  takeProfit?: number;
  exitPrice?: number;
  quantity?: number;
  profitLoss?: number;
  notes?: string;
  tags?: string[];
  tradeTime?: string;
}

export interface TradeLogQuery {
  stockCode?: string;
  tags?: string[];
  startDate?: string;
  endDate?: string;
  page?: number;
  pageSize?: number;
}

export interface TradeLogListResponse {
  items: TradeLog[];
  total: number;
  page: number;
  pageSize: number;
}

File: backend/app/schemas/log.py
"""
Trade log schemas for API request/response
"""

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import PatternEnum, PeriodEnum


class TradeLogCreate(BaseModel):
    """Trade log creation request"""
    stockCode: str = Field(..., max_length=10)
    stockName: str | None = Field(None, max_length=50)
    period: PeriodEnum
    patternType: PatternEnum | None = None
    entryPrice: Decimal = Field(..., gt=0)
    stopLoss: Decimal | None = None
    takeProfit: Decimal | None = None
    exitPrice: Decimal | None = None
    quantity: int = Field(..., gt=0)
    profitLoss: Decimal | None = None
    notes: str | None = None
    tags: list[str] | None = None
    tradeTime: datetime | None = None


class TradeLogUpdate(BaseModel):
    """Trade log update request"""
    stockCode: str | None = None
    stockName: str | None = None
    period: PeriodEnum | None = None
    patternType: PatternEnum | None = None
    entryPrice: Decimal | None = None
    stopLoss: Decimal | None = None
    takeProfit: Decimal | None = None
    exitPrice: Decimal | None = None
    quantity: int | None = None
    profitLoss: Decimal | None = None
    notes: str | None = None
    tags: list[str] | None = None
    tradeTime: datetime | None = None


class TradeLogResponse(BaseModel):
    """Trade log response"""
    id: UUID
    stockCode: str
    stockName: str | None = None
    period: PeriodEnum
    patternType: PatternEnum | None = None
    entryPrice: Decimal
    stopLoss: Decimal | None = None
    takeProfit: Decimal | None = None
    exitPrice: Decimal | None = None
    quantity: int
    profitLoss: Decimal | None = None
    notes: str | None = None
    tags: list[str] | None = None
    tradeTime: datetime | None = None
    createdAt: datetime
    updatedAt: datetime

    class Config:
        from_attributes = True


class TradeLogQuery(BaseModel):
    """Trade log query parameters"""
    stockCode: str | None = None
    tags: list[str] | None = None
    startDate: datetime | None = None
    endDate: datetime | None = None


class TradeLogListResponse(BaseModel):
    """Paginated trade log list response"""
    items: list[TradeLogResponse]
    total: int
    page: int
    pageSize: int
